selfservice/strategy/saml/handler.go

```diff
@@ -0,0 +1,374 @@
+
+	// We check if the metadata file is provided
+	if providerConfig.IDPInformation["idp_metadata_url"] != "" {
+		// The metadata file is provided
+		metadataURL := providerConfig.IDPInformation["idp_metadata_url"]
+
+		metadataBuffer, err := fetcher.NewFetcher().Fetch(metadataURL)
+		if err != nil {
+			return herodot.ErrNotFound.WithTrace(err)
+		}
+
+		metadata, err := ioutil.ReadAll(metadataBuffer)
+		if err != nil {
+			return herodot.ErrInternalServerError.WithTrace(err)
+		}
+
+		idpMetadata, err = samlsp.ParseMetadata(metadata)
+		if err != nil {
+			return ErrInvalidSAMLMetadataError.WithTrace(err)
+		}
+
+	} else {
+		// The metadata file is not provided
+		// So were are creating a minimalist IDP metadata based on what is provided by the user on the config file
+		entityIDURL, err := url.Parse(providerConfig.IDPInformation["idp_entity_id"])
+		if err != nil {
+			return herodot.ErrNotFound.WithTrace(err)
+		}
+
+		// The IDP SSO URL
+		IDPSSOURL, err := url.Parse(providerConfig.IDPInformation["idp_sso_url"])
+		if err != nil {
+			return herodot.ErrNotFound.WithTrace(err)
+		}
+
+		// The certificate of the IDP
+		certificateBuffer, err := fetcher.NewFetcher().Fetch(providerConfig.IDPInformation["idp_certificate_path"])
+		if err != nil {
+			return herodot.ErrNotFound.WithTrace(err)
+		}
+
+		certificate, err := ioutil.ReadAll(certificateBuffer)
+		if err != nil {
+			return herodot.ErrInternalServerError.WithTrace(err)
+		}
+
+		// We parse it into a x509.Certificate object
+		IDPCertificate, err := MustParseCertificate(certificate)
+		if err != nil {
+			return ErrInvalidCertificateError.WithTrace(err)
+		}
+
+		// Because the metadata file is not provided, we need to simulate an IDP to create artificial metadata from the data entered in the conf file
+		tempIDP := samlidp.IdentityProvider{
+			Certificate: IDPCertificate,
+			MetadataURL: *entityIDURL,
+			SSOURL:      *IDPSSOURL,
+		}
+
+		// Now we assign our reconstructed metadata to our SP
+		idpMetadata = tempIDP.Metadata()
+	}
+
+	// The main URL
+	rootURL, err := url.Parse(config.SelfServiceBrowserDefaultReturnTo(ctx).String())
+	if err != nil {
+		return herodot.ErrNotFound.WithTrace(err)
+	}
+
+	// Here we create a MiddleWare to transform Kratos into a Service Provider
+	samlMiddleWare, err := samlsp.New(samlsp.Options{
+		URL:         *rootURL,
+		Key:         keyPair.PrivateKey.(*rsa.PrivateKey),
+		Certificate: keyPair.Leaf,
+		IDPMetadata: idpMetadata,
+		SignRequest: true,
+		// We have to replace the ContinuityCookie by using RelayState. We will pass the SessionID (uuid) of Kratos through RelayState
+		RelayStateFunc: func(w http.ResponseWriter, r *http.Request) string {
+			ctx := r.Context()
+			cipheredCookie, ok := ctx.Value(ory_kratos_continuity{}).(string)
+			if !ok {
+				_, err := w.Write([]byte("No SessionID in current context"))
+				if err != nil {
+					h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+				}
+				return ""
```

Comment:
Check what the library does when this is empty? Probably need to test manually because not clear from the code path.